PhysicsTools/PatAlgos/python/slimming/customizeMiniAOD_HcalFixLegacy2016.py

```diff
@@ -0,0 +1,134 @@
+import FWCore.ParameterSet.Config as cms
+
+from PhysicsTools.PatAlgos.tools.helpers import MassSearchReplaceAnyInputTagVisitor, addKeepStatement
+from PhysicsTools.PatAlgos.tools.helpers import getPatAlgosToolsTask
+
+def loadJetMETBTag(process):
+
+    task = getPatAlgosToolsTask(process)
+
+    import RecoJets.Configuration.RecoPFJets_cff
+    process.ak4PFJetsCHS = RecoJets.Configuration.RecoPFJets_cff.ak4PFJetsCHS.clone()
+    task.add(process.ak4PFJetsCHS)
+    # need also the non-CHS ones as they are used to seed taus
+    process.ak4PFJets = RecoJets.Configuration.RecoPFJets_cff.ak4PFJets.clone()
+    task.add(process.ak4PFJets)
+    process.ak8PFJetsCHS = RecoJets.Configuration.RecoPFJets_cff.ak8PFJetsCHS.clone()
+    task.add(process.ak8PFJetsCHS)
```

Comment:
please follow up on this comment

Reply:
your suggestion doesn't seem to work out of the box. I get a:
`ValueError: Trying to override definition of ak4PFJetsPuppi while it is used by the task patAlgosToolsTask`
focusing on the other comments for now

Reply:
```customising the process with customizeAll from PhysicsTools/PatAlgos/slimming/customizeMiniAOD_HcalFixLegacy2016
Traceback (most recent call last):
  File "/cvmfs/cms-ib.cern.ch/nweek-02521/slc6_amd64_gcc630/cms/cmssw/CMSSW_10_2_X_2018-04-22-1100/bin/slc6_amd64_gcc630/cmsDriver.py", line 55, in <module>
    run()
  File "/cvmfs/cms-ib.cern.ch/nweek-02521/slc6_amd64_gcc630/cms/cmssw/CMSSW_10_2_X_2018-04-22-1100/bin/slc6_amd64_gcc630/cmsDriver.py", line 27, in run
    configBuilder.prepare()
  File "/cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_2_X_2018-04-23-2300/python/Configuration/Applications/ConfigBuilder.py", line 2252, in prepare
    self.pythonCfgCode += self.addCustomise(1)
  File "/cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_2_X_2018-04-23-2300/python/Configuration/Applications/ConfigBuilder.py", line 881, in addCustomise
    self.process=getattr(package,fcn)(self.process)
  File "/afs/cern.ch/work/d/deguio/HCAL/ChannelCleaning/CMSSW_10_2_X_2018-04-23-2300_rebaseOfOfficialRecipe/python/PhysicsTools/PatAlgos/slimming/customizeMiniAOD_HcalFixLegacy2016.py", line 102, in customizeAll
    loadJetMETBTag(process)
  File "/afs/cern.ch/work/d/deguio/HCAL/ChannelCleaning/CMSSW_10_2_X_2018-04-23-2300_rebaseOfOfficialRecipe/python/PhysicsTools/PatAlgos/slimming/customizeMiniAOD_HcalFixLegacy2016.py", line 10, in loadJetMETBTag
    process.load("RecoJets.Configuration.RecoPFJets_cff")
  File "/cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_2_X_2018-04-23-2300/python/FWCore/ParameterSet/Config.py", line 583, in load
    self.extend(sys.modules[moduleName])
  File "/cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_2_X_2018-04-23-2300/python/FWCore/ParameterSet/Config.py", line 606, in extend
    self.__setattr__(name,item)
  File "/cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_2_X_2018-04-23-2300/python/FWCore/ParameterSet/Config.py", line 375, in __setattr__
    raise ValueError(msg1+s.label_()+msg2)
ValueError: Trying to override definition of ak4PFJetsPuppi while it is used by the task patAlgosToolsTask
 new object defined in: /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_2_X_2018-04-23-2300/python/RecoJets/JetProducers/ak4PFJets_cfi.py
 existing object defined in: /afs/cern.ch/work/d/deguio/HCAL/ChannelCleaning/CMSSW_10_2_X_2018-04-23-2300_rebaseOfOfficialRecipe/python/PhysicsTools/PatAlgos/slimming/applySubstructure_cff.py```

Reply:
The error is because ak4PFJetsPuppi is defined in multiple places.

This is more messy then. 
The error also exposes the dangers of loading the reco pieces via customization: logically, they should be loaded first, before the miniAOD comes.


I think that we need the rhos (currently in recoPFJetsTask) to run.
Is it obvious that none of the other 

So, they perhaps should  be added here explicitly.
A slightly cleaner solution involves changing RecoJets/Configuration/python/RecoPFJets_cff.py : to get the rhos, define fixedGridRhos_cff with the 5 fixedGridRhos currently defined in RecoPFJets_cff and load it there and here as well.

A

Reply:
On a second thought, I think that it's worth the effort to load the reco pieces in the same order as they are normally added in production where miniAOD and RECO are running. 
This means moving all the reco definitions/task inclusion up the chain, before the miniAOD is defined.